net: use std::chrono throughout maxOutbound logic

fanquake opened the pull request:
Switch to using `std::chrono` types for the max outbound related logic.
Removes some unnecessary code from init.

fanquake committed:
net: remove SetMaxOutboundTarget

This has been unused since f3552da81393a8e78ce3e2afed0b9c9d1ff5cee0.

fanquake committed:
net: remove SetMaxOutboundTimeframe

This was introduced in 872fee3fccc8b33b9af0a401b5f85ac5504b57eb and it's unclear
if it's ever been used.

fanquake committed:
net: remove nMaxOutboundTimeframe from connection options

It's not actually possible to change this value, so remove the
indirection of it being a conn option.

DEFAULT_MAX_UPLOAD_TIMEFRAME is a compile time constant.

fanquake committed:
init: set nMaxOutboundLimit connection option directly

DEFAULT_MAX_UPLOAD_TARGET is a compile time constant.

fanquake committed:
net: use std::chrono throughout maxOutbound logic

practicalswift commented:
Concept ACK

theStack reviewed (commented):
Concept ACK
Note that the functions `SetMaxOutbound{Target,Timeframe}` are also used in the fuzz test `test/fuzz/connman.cpp`, hence the uses should also be removed there.

fanquake commented:
> Note that the functions SetMaxOutbound{Target,Timeframe} are also used in the fuzz test test/fuzz/connman.cpp, hence the uses should also be removed there.

Thanks. I've dropped them from the fuzz tests now.

jnewbery commented:
Concept ACK

fanquake commented:
Rebased for #20408 and fixed an oversight in the fuzz changes.

theStack commented on src/net.cpp:
```diff
@@ -2837,24 +2837,24 @@ uint64_t CConnman::GetMaxOutboundTarget()
-    return (cycleEndTime < now) ? 0 : cycleEndTime - GetTime();
+    std::chrono::seconds cycleEndTime = nMaxOutboundCycleStartTime + nMaxOutboundTimeframe;
+    auto now = GetTime<std::chrono::seconds>();
+    return (cycleEndTime < now) ? std::chrono::seconds::zero() : cycleEndTime - GetTime<std::chrono::seconds>();
```
Could reuse `now` here instead of calling `GetTime()` again:
```return (cycleEndTime < now) ? std::chrono::seconds::zero() : cycleEndTime - now;```

fanquake replied:
Thanks. I was going to leave this as a straight refactor, but have  taken your suggestion.

MarcoFalke commented on src/net.cpp:
```diff
@@ -2818,48 +2818,30 @@ void CConnman::RecordBytesSent(uint64_t bytes)
-    nMaxOutboundTimeframe = timeframe;
+    std::chrono::seconds cycleEndTime = nMaxOutboundCycleStartTime + nMaxOutboundTimeframe;
+    auto now = GetTime<std::chrono::seconds>();
+    return (cycleEndTime < now) ? std::chrono::seconds::zero() : cycleEndTime - now;
```
I think 0 can be written shorter with chrono literals?

fanquake replied:
Done

jnewbery commented on src/net.h:
```diff
@@ -361,13 +361,8 @@ class CConnman
-    //!set the timeframe for the max outbound target
-    void SetMaxOutboundTimeframe(uint64_t timeframe);
-    uint64_t GetMaxOutboundTimeframe();
+    std::chrono::seconds GetMaxOutboundTimeframe();
```
It's a good idea to use `std::chrono` internally in net, but I think you can leave these interface functions `GetMaxOutboundTimeframe()` and `GetMaxOutboundTimeLeftInCycle()` the same and continue to just send a count of seconds to the caller. The only place these are used is in rpc/net, where the count of seconds is returned directly to the user.

fanquake replied:
> The only place these are used is in rpc/net, where the count of seconds is returned directly to the user.

`GetMaxOutboundTimeLeftInCycle` is called in `CConnman::OutboundTargetReached`.

jnewbery replied:
Ah, of course.

jnewbery reviewed (commented):
Concept ACK, especially cleaning up init.

Longer term, it'd be nice to move initialization of some of these CConnman parameters into the constructor and make them const. That clarifies the initialization and means that they don't need to be guarded by a mutex.

fanquake commented:
I've rebased this, cherry-picked in fa11110bff6288f63e0c487e2e4b4079fb0f4569 from 20602 and updated some std::chrono usage to use literals as suggested.

fanquake commented on src/net.h:
```diff
@@ -209,7 +209,7 @@ class CConnman
         unsigned int nSendBufferMaxSize = 0;
         unsigned int nReceiveFloodSize = 0;
-        uint64_t nMaxOutboundTimeframe = 0;
+        std::chrono::seconds nMaxOutboundTimeframe{0};
```
I forgot to do this in the last push, but should have addressed it now.

jonatack commented on src/net.cpp:
```diff
@@ -2806,8 +2806,8 @@ void CConnman::RecordBytesSent(uint64_t bytes)
 
-    uint64_t now = GetTime();
-    if (nMaxOutboundCycleStartTime + nMaxOutboundTimeframe < now)
+    auto now = GetTime<std::chrono::seconds>();
```
832fc10e9e45abf9 const and maybe explicit typing
```suggestion
    const std::chrono::seconds now{GetTime<std::chrono::seconds>()};
```

fanquake replied:
Having `std::chrono::seconds` twice on the same line is pretty verbose. I'd like to think we can just use `auto` here because the type is clear from the call to GetTime<>().

jonatack replied:
Yes, the reason for the comment here was constness.

jonatack replied:
Note that existing code in the same file specifies the type explicitly, line 1646: `const std::chrono::seconds now = GetTime<std::chrono::seconds>();`

jonatack reviewed (commented):
ACK 832fc10e9e45abf92a4f04012c918156d92a3649

with some optional style ideas below

jnewbery commented on src/net.h:
```diff
@@ -73,9 +73,9 @@ static const bool DEFAULT_UPNP = false;
+static constexpr uint64_t DEFAULT_MAX_UPLOAD_TARGET = 0;
 /** The default timeframe for -maxuploadtarget. 1 day. */
-static const uint64_t MAX_UPLOAD_TIMEFRAME = 60 * 60 * 24;
+static constexpr std::chrono::seconds MAX_UPLOAD_TIMEFRAME{60 * 60 * 24};
```
Since this is only used internally in net, it could now be moved to the implementation cpp file to avoid exporting this symbol to other translation units.

fanquake replied:
Done in the latest push.

jnewbery reviewed (commented):
utACK 832fc10e9e

In case you touch this branch again: In commit _net: remove nMaxOutboundTimeframe from connection options_ I think you need to change "DEFAULT_MAX_UPLOAD_TARGET is a compile time constant." to "DEFAULT_MAX_UPLOAD_TIMEFRAME is a compile time constant."

fanquake commented:
Rebased & addressed some comments.

> In case you touch this branch again:

I've fixed the commit message.

maflcko reviewed (approved):
review ACK 0475c8ba4d10dce79092361bc4c23c11dadba39a 🎭

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 0475c8ba4d10dce79092361bc4c23c11dadba39a 🎭
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjthwwAzVabd9gXhamR9ER43crYv/VUAp95aBP6//eaBjC5dwfvq8UzeJtoP9aF
NiVZ14+MDWyLQyUBZF+DYU3veWMHYgaIGSpOQ+ZlgVYCva2ZmkhqnlCk4de4jjcL
1WMJV9zxauGhITHd+AVqy2SyKd/6ljAMWn3v89KYqDApw7zzEIe5Jpbfq7f7kbbh
6d9X1XqQoSCNS+SwCouEu192Y9f+44jAo08FupUshCHMUMV+MZs87Hs3rD+N4DsX
95AZV1gO48HgDldPvUC7LZIieTIJFBpI4Y7dr/Re186h1csWc++8FbXOTLN12gg2
+KR4/XgqYVBNKBhv/X60Hc113MReXc3gbGqXtFMu1vVXRJnEXM5E/bk9qxYzsdFX
n5D97DMHOotQh1W61f7Mqlqk9N+qP28P2BfJwxamyVC1DQ4PiLvlY79kxm3UkWGx
BzFD1qqn7M67Twm2z1SO0nBgQZCXMPsmFI0dR+abUT4o5jUXMIovVef26reBTS8v
ReieyZiR
=+TOK
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `963300e8377971b0f423721ee34e6ab50d44d071f41bb4f7f3598935d178a88d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108963300e8377971b0f423721ee34e6ab50d44d071f41bb4f7f3598935d178a88df0106e44748041798fd03fa5c9f29012e8c408fff01080a2974c8ab21289ed0f1a705d5a37b108f1045fd8763ff008f39be97fe1142f3f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010a9f29c6d8a805e7d50d6001a8a50731f08f1045fd8763ff008b2e2b80fb0f6936f0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c5f24f9c626f7a353c60e15c2715b8f808f1045fd8763ff008aa1bdcd0cc1685330083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010d9266d9d4021f3458b09697576c31bf708f120c23bb68f4a9b84f0c953bcf837bd9f4eb56d1f42805930f5ac78d1775c63800208f1045fd8763ff008aa6e5950aa9142320083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jnewbery commented:
utACK 0475c8ba4d10dce79092361bc4c23c11dadba39a
